sway-core/src/metadata.rs

```diff
@@ -61,6 +63,33 @@ impl MetadataManager {
         })
     }
 
+    pub(crate) fn md_to_decl_index(
+        &mut self,
+        context: &Context,
+        md_idx: Option<MetadataIndex>,
+    ) -> Option<usize> {
+        Self::for_each_md_idx(context, md_idx, |md_idx| {
+            self.md_decl_index_cache.get(&md_idx).cloned().or_else(|| {
+                // Create a new decl index and save it in the cache
+                md_idx
+                    .get_content(context)
+                    .unwrap_struct("decl_index", 2)
+                    .and_then(|fields| {
+                        // First check if the read metadata corresponds to Some(_) or None
+                        let is_index_some = fields[0].unwrap_integer().map(|index| index != 0)?;
+                        let index = if is_index_some {
+                            Some(fields[1].unwrap_integer().map(|index| index as usize)?)
+                        } else {
+                            None
+                        };
+                        self.md_decl_index_cache.insert(md_idx, index);
```

Comment:
I ended up doing some juggling to read `Option<>` value from the metadata like this, curious about whether there is a more idiomatic way to replace this stuff.